site/app/libraries/database/PostgresqlDatabaseQueries.php

```diff
@@ -825,5 +829,305 @@ public function updateLateDays($user_id, $timestamp, $days, $csv_option=null) {
+                      gc_id,
+                      gd_id
+                    FROM gradeable_component_mark_data
+                    GROUP BY gc_id, gd_id
+                  ) AS gcmd ON gcmd.gc_id=in_gcd.gc_id AND gcmd.gd_id=in_gcd.gd_id
+                GROUP BY in_gcd.gd_id
+              ) AS gcd ON gcd.gd_id=gd.gd_id
+              LEFT JOIN (
+                SELECT * 
+                FROM electronic_gradeable_version
+              ) AS egv ON (egv.team_id=gd.gd_team_id OR egv.user_id=gd.gd_user_id) AND egv.g_id=gd.g_id
+              LEFT JOIN (
+                SELECT
+                  array_agg(in_egd.g_version) AS array_version,
+                  array_agg(in_egd.autograding_non_hidden_non_extra_credit) AS array_non_hidden_non_extra_credit,
+                  array_agg(in_egd.autograding_non_hidden_extra_credit) AS array_non_hidden_extra_credit,
+                  array_agg(in_egd.autograding_hidden_non_extra_credit) AS array_hidden_non_extra_credit,
+                  array_agg(in_egd.autograding_hidden_extra_credit) AS array_hidden_extra_credit,
+                  array_agg(in_egd.submission_time) AS array_submission_time,
+                  array_agg(in_egd.autograding_complete) AS array_autograding_complete,
+                  g_id,
+                  user_id,
+                  team_id
+                FROM electronic_gradeable_data as in_egd
+                GROUP BY g_id, user_id, team_id
+              ) AS egd ON (egd.team_id=gd.gd_team_id OR egd.user_id=gd.gd_user_id) AND egd.g_id=gd.g_id
+            WHERE gd.g_id=? AND ((gd.gd_user_id IN (?)) OR (gd.gd_team_id IN (?)) OR ?)";
+
+        $this->course_db->query($query, array(
+            $gradeable->getId(),
+            implode(',', $users),
+            implode(',', $teams),
+            $this->course_db->convertBoolean($all)));
+
+        $graded_gradeables = [];
+
+        $db_rows = $this->course_db->rows();
+
+        $user_ids = [];
+        $team_ids = [];
+
+        // Cache the graders / submitters
+        foreach($db_rows as $db_row) {
+            $user_ids = array_merge($user_ids, $this->course_db->fromDatabaseToPHPArray($db_row['array_grader_id']));
+
+            $s_user = $db_row['user_id'];
+            $s_team = $db_row['team_id'];
+            if($s_user !== null) {
+                $user_ids[] = $s_user;
+            }
+            if($s_team !== null) {
+                $team_ids[] = $s_team;
+            }
+        }
+
+        // Remove null / duplicate values being sneaky in our arrays
+        $filter_null = function($val) { return $val !== null; };
+        $user_ids = array_filter(array_unique($user_ids), $filter_null);
+        $team_ids = array_filter(array_unique($team_ids), $filter_null);
+
+        $users = $this->getUsersById($user_ids);
+        $teams = $this->getTeamsById($team_ids);
+
+        // decode the gradeable data
+        foreach ($db_rows as $db_row) {
```

Comment:
It should be noted that the reason the user and team information was being joined was because it allowed us to write one query, which while yes, did get some duplicate information per row, meant that we could use a DatabaseIterator to go through the data row by row when we needed the info. This is drastically more efficient in both time spent and memory used as we largely don't need all of the information at one time, but rather row by row.